elfi/methods/utils.py

```diff
@@ -184,29 +184,70 @@ def logpdf(cls, x, means, cov=1, weights=None):
         return np.log(cls.pdf(x, means=means, cov=cov, weights=weights))
 
     @classmethod
-    def rvs(cls, means, cov=1, weights=None, size=1, random_state=None):
+    def rvs(cls, means, cov=1, weights=None, size=1, prior_logpdf=None, random_state=None):
         """Draw random variates from the distribution.
 
         Parameters
         ----------
         means : array_like
             means of the Gaussian mixture components
-        weights : array_like
-            1d array of weights of the gaussian mixture components
-        cov : array_like
+        cov : array_like, optional
             a shared covariance matrix for the mixture components
-        size : int or tuple
-        random_state : np.random.RandomState or None
+        weights : array_like, optional
+            1d array of weights of the gaussian mixture components
+        size : int or tuple, optional
+        prior_logpdf : callable, optional
+            Can be used to check validity of random variable.
+        random_state : np.random.RandomState, optional
 
         """
-        means, weights = cls._normalize_params(means, weights)
         random_state = random_state or np.random
+        means, weights = cls._normalize_params(means, weights)
 
-        inds = random_state.choice(len(means), size=size, p=weights)
-        rvs = means[inds]
-        perturb = ss.multivariate_normal.rvs(
-            mean=means[0] * 0, cov=cov, random_state=random_state, size=size)
-        return rvs + perturb
+        # additional complexity due to supporting scipy-like ambiguity of arguments
+        if size is None:
+            size = means.shape[0]
```

Comment:
I don't quite understand this line. Why would we want as many simulations as there are mean vectors if `size` is `None`?

Reply:
Ok, I have to say that I'm not exactly sure what it's supposed to return when `size=None`. The code is written so that the tests pass. :D

Honestly I suppose we could drop support for `size=None` (and hence scipy-like behaviour?), since it's unlikely to be used, and complicates the code.

Reply:
So assuming e.g. 2d vector outputs, `size=1` would return `[[1, 2]]` while `size=None` returns `[1,2]`. Basically with size `None` (should be btw. changed to default to follow scipy like behaviour) the output is not wrapped to a containing array.

With 1d output, `size=None` returns `numpy` scalars, and `size=1` returns the scalar in an array.

Reply:
Hmm...? It seems to me that the `size` in `scipy.stats` is actually `shape`?

`ss.norm.rvs(loc=np.random.rand(1,2), size=1).shape` raises ValueError
`ss.norm.rvs(loc=np.random.rand(1,2), size=2).shape` gives (2,)
`ss.norm.rvs(loc=np.random.rand(1,2), size=None).shape` gives (2,)
`ss.norm.rvs(loc=np.random.rand(1,2), size=(1,2)).shape` gives (1,2)

Using `loc=np.random.rand(2)` gives the same shapes.

`ss.norm.rvs(loc=np.random.rand(1), size=1).shape` gives (1,)
`ss.norm.rvs(loc=np.random.rand(1), size=2).shape` gives (2,)
`ss.norm.rvs(loc=np.random.rand(1), size=None).shape` gives ()
`ss.norm.rvs(loc=np.random.rand(1), size=(1,2)).shape` gives (1,2)

Reply:
Yes, with `ss.norm`, but this uses `ss.multivariate_normal`

Reply:
Currently, yes, but we should change it to ~arbitrary distribution.